### PR TITLE
Optimize count_distinct.size

### LINKED ISSUE #5325: Optimize Accumulator `size` function performance (fix regression on clickbench)
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
During regression benchmarks it was found that DISTINCT queries has a performance drop. The analysis showed the `size` function implementation for `DistinctCountAccumulator` is inefficient. 

**Describe the solution you'd like**
Need to improve size function or the way how number of bytes is collected

**Describe alternatives you've considered**
None

**Additional context**
Analysis details https://github.com/apache/arrow-datafusion/issues/5313 
Original benchmarks ticket https://github.com/apache/arrow-datafusion/issues/5276

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #5325 .

# Rationale for this change
Fixing performance drop for `.size` function. Drop was found during performance benchmarks
During regression the query with `count(distinct )` took 100s in optimized build, now it takes 5s.
 

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by comphead** [comment]
@alamb @crepererum please check this PR whenever you have time

File: datafusion/physical-expr/src/aggregate/count_distinct.rs [hunk 1]
```diff
@@ -216,23 +216,19 @@ impl Accumulator for DistinctCountAccumulator {
     }
 
     fn size(&self) -> usize {
+        // temporarily calculating the size approximately, taking first batch size * number of batches
+        // such approach has some inaccuracy for variable length values, like strings.
```
Comment (crepererum) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: This basically remove proper memory accounting for this operation and for strings will likely be very wrong. I would rather see a proper cached size accounting here.

Comment (comphead) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: Right, as we agreed in https://github.com/apache/arrow-datafusion/issues/5325#issuecomment-1441767680
We need a fix the benchmark and later @alamb can think how to deal with variable length data.

Comment (crepererum) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: This is clearly a conflict of interests: you wanna fix the benchmark, I wanna have proper memory accounting. We likely can have both on the long run, but due to limited development resources, we cannot have the ideal solution right now. 

I'll leave it to the project managers (e.g. @alamb) to decide what's more pressing.

Comment (jychen7) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: > Instead add some code that checks "if is ScalarType::Int8, UInt8, etc then size = size[0]*vec.len()"

https://github.com/apache/arrow-datafusion/issues/5325#issuecomment-1435672826

are we missing this check in this PR?

Comment (alamb) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: > are we missing this conditionall check in this PR? so we still have accurate size (slow for now) for variable data and accurate size (fast) for fixed lenth data

This is the middle path I would suggest: keep the slow but accurate accounting for variable length data (aka strings) and add a fast path for fixed length sizes (what is in the benchmark)

I believe the the additional overhead of doing accurate size accounting for string values is a relatively smaller amount of the overall time compared to fixed size types. Making `count distinct` with a large number of string values fast is likely going to take a more sophisticated approach to this query in general.

Comment (comphead) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: @alamb @crepererum Let me remind you again another possible temp solution - The fastest and most painless fix to fix benchmark is to increase batch number, and .size function will be called less often.

Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:220]: The overhead for an accurate answer is still enormous for variable length types and gets worse with increasing table sizes, as the `size` gets slower with more distinct values in the aggregation, and is called for every update (so a `O(n^2)` operation).

Comment (comphead) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs]: @alamb @crepererum @jychen7 @Dandandan Amended to get approx_size for primitives only

File: datafusion/physical-expr/src/aggregate/count_distinct.rs [hunk 2]
```diff
@@ -149,6 +149,43 @@ impl DistinctCountAccumulator {
             self.update(&row_values)
         })
     }
+
+    // calculating the size approximately, taking first batch size * number of batches
+    // approx_size has some inaccuracy for variable length values, like strings.
+    fn approx_size(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:155]: So we could change the name to e.g. `fixed_size`

File: datafusion/physical-expr/src/aggregate/count_distinct.rs [hunk 3]
```diff
@@ -149,6 +149,43 @@ impl DistinctCountAccumulator {
...
+
+    // calculating the size for fixed length values, taking first batch size * number of batches
+    // This method is faster than .full_size(), however it is not suitable for variable length values like strings or complex types
+    fn fixed_size(&self) -> usize {
+        std::mem::size_of_val(self)
+            + (std::mem::size_of::<DistinctScalarValues>() * self.values.capacity())
+            + self
+                .values
+                .iter()
+                .next()
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:161]: 👍  that is nice

File: datafusion/physical-expr/src/aggregate/count_distinct.rs [hunk 4]
```diff
@@ -216,23 +253,27 @@ impl Accumulator for DistinctCountAccumulator {
...
+            | DataType::Float32
+            | DataType::Float64
+            | DataType::Int16
+            | DataType::Int32
+            | DataType::Int64
+            | DataType::Int8
+            | DataType::Time32(_)
+            | DataType::Time64(_)
+            | DataType::Null
+            | DataType::Timestamp(_, _)
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:270]: I think this is missing some types like `Interval`  and `Duration`

Perhaps you could check instead if `DataType::primitive_width` returned `Some(.)` 🤔 

https://docs.rs/arrow/34.0.0/arrow/datatypes/enum.DataType.html#method.primitive_width

**Review by alamb — APPROVED** [review]
This is looking good @comphead  -- I had one more suggestion about how to test which types were primitive but I also think this code is an improvement over what is on master so it could be merged as well.

Thank you so much

File: datafusion/physical-expr/src/aggregate/count_distinct.rs (hunk 4 above)
Comment (comphead) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs]: Done, seems in new arrow-rs @tustvold made some work for us, and introduced `Datatype::is_primitive`

File: datafusion/physical-expr/src/aggregate/count_distinct.rs [hunk 5]
```diff
@@ -216,23 +253,11 @@ impl Accumulator for DistinctCountAccumulator {
...
-                .sum::<usize>()
-            + (std::mem::size_of::<DataType>() * self.state_data_types.capacity())
-            + self
-                .state_data_types
-                .iter()
-                .map(|dt| dt.size() - std::mem::size_of_val(dt))
-                .sum::<usize>()
-            + self.count_data_type.size()
-            - std::mem::size_of_val(&self.count_data_type)
+        if self.count_data_type.is_primitive() {
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:256]: 👍

**Review by alamb — APPROVED** [review]
Thanks @comphead  -- this looks great

**Pull request merged**